app/models/api/translation_memory.php

```diff
@@ -32,21 +33,51 @@
 // We loop through all repositories and merge the results
 foreach ($repositories as $repository) {
     $source_strings = Utils::getRepoStrings($request->parameters[3], $repository);
-    $target_strings = Utils::getRepoStrings($request->parameters[4], $repository);
 
     foreach ($terms as $word) {
         $search->setRegexSearchTerms($word);
         $source_strings = preg_grep($search->getRegex(), $source_strings);
     }
 
-    $source_strings_merged = array_merge($source_strings, $source_strings_merged);
-    $target_strings_merged = array_merge($target_strings, $target_strings_merged);
+    /*
+        If we don't have any match for a repo, no need to do heavy calculation,
+        just skip to the next repo.
+    */
+    if (empty($source_strings)) {
+        continue;
+    }
+
+    /*
+        We are only interested in target strings with keys in common with our
+        source strings. Not sending noise to getTranslationMemoryResults() has
+        a major performance and memory impact.
+    */
+    $target_strings = array_intersect_key(
+        Utils::getRepoStrings($request->parameters[4], $repository),
+        $source_strings
+    );
+
+    /*
+        We are not interested in keeping duplicate strings that have
+        different keys because this API does not take into account the
+        frequency of matches but the similarity of the strings.
+    */
+    $target_strings = array_unique($target_strings);
```

Comment:
I might be lost, but what's the scenario here? ```$target_strings``` is set with new values for each repository, given that keys are in the form `browser/chrome/browser/aboutPrivateBrowsing.dtd:aboutPrivateBrowsing.info.cookies` in which cases would we have duplicated items?